### PR TITLE
Feature/concurrent dcopy

### LINKED ISSUE #81: Option to copy two or more files simultaneous
hi, is it possible to add an option to copy two or more files simultaneous?
I think this would be an amazing performance boost on the most systems.

What do you think?

### PR DESCRIPTION
**Pull request by otiai10**
Fix #81 
Fix #120

### CODE REVIEW

**Comment by otiai10** [comment]
@ycombinator Please take a look and comment. Please try this version on your local as well if possible.

**Comment by otiai10** [comment]
@ycombinator any comment on this?

**Comment by ycombinator** [comment]
@otiai10 I have not yet had time to look at this.  My plan is to look at it tomorrow (Friday). I took a quick glance at it and noticed one major difference from my implementation in https://github.com/otiai10/copy/pull/120 is that in this one you are only concurrently copying directories.  Whereas, in 120, all files were being copied concurrently.  Any reason for this difference?

**Comment by otiai10** [comment]
When we call `copy.Copy`, the parameter must be **an entry**. This is the key fact.
Let's name it **a root entry**.

* If the root entry is `a file`, no need to be concurrent.
* If the root entry is `a pipe`, `a symlink`, do not copy concurrently.
* If the root entry is `a directory`, copy concurrently.

Thus I did like this.

But I found a critical bug of this implementation in another context, thus I'll close this.

**Comment by otiai10** [comment]
Fixed

File: README.md [hunk 1]
```diff
@@ -86,6 +86,19 @@ type Options struct {
 	// If given, copy.Copy refers to this fs.FS instead of the OS filesystem.
 	// e.g., You can use embed.FS to copy files from embedded filesystem.
 	FS fs.FS
+
+	// NumOfWorkers represents the number of workers used for
+	// concurrent copying contents of directories.
+	// If 0 or 1, it does not use goroutine for copying directories.
+	// Please refer to https://pkg.go.dev/golang.org/x/sync/semaphore for more details.
+	NumOfWorkers int64
```
Comment (ycombinator) [inline README.md:94]: Any reason not to make this just an `int`?

Comment (otiai10) [reply README.md:94]: According to semaphore.NewWeighted

**Review by ycombinator — APPROVED** [review]
LGTM - thank you!

I tested this PR with https://github.com/elastic/elastic-agent/pull/3212#issue-1842215062 and updated the performance results in the table. There is definitely a speed improvement!

**Comment by otiai10** [comment]
@shiwork That would be great if you can leave some comment on this PR :bow:

**Comment by ycombinator** [comment]
Hi @otiai10, any plans to merge this PR? We are trying to use this improvement in [Elastic Agent](https://www.elastic.co/elastic-agent) during installation (see https://github.com/elastic/elastic-agent/pull/3212).  Thanks!

**Comment by otiai10** [comment]
@ycombinator Thank you for pinging. I'll get back to you soon 🙇

**Comment by otiai10** [comment]
@ycombinator Released under v1.14.0

**Pull request merged**